execution/storage/pool.go

```diff
@@ -27,37 +28,56 @@ func NewSelectorPool(queryable storage.Queryable) *SelectorPool {
 	}
 }
 
-func (p *SelectorPool) GetSelector(mint, maxt int64, matchers []*labels.Matcher) SeriesSelector {
-	key := hashMatchers(matchers, mint, maxt)
+func (p *SelectorPool) GetSelector(mint, maxt, step int64, matchers []*labels.Matcher, hints *storage.SelectHints) SeriesSelector {
+	key := hashMatchers(matchers, mint, maxt, hints)
 	if _, ok := p.selectors[key]; !ok {
-		p.selectors[key] = newSeriesSelector(p.queryable, mint, maxt, matchers)
+		p.selectors[key] = newSeriesSelector(p.queryable, mint, maxt, step, matchers, hints)
 	}
 	return p.selectors[key]
 }
 
-func (p *SelectorPool) GetFilteredSelector(mint, maxt int64, matchers, filters []*labels.Matcher) SeriesSelector {
-	key := hashMatchers(matchers, mint, maxt)
+func (p *SelectorPool) GetFilteredSelector(mint, maxt, step int64, matchers, filters []*labels.Matcher, hints *storage.SelectHints) SeriesSelector {
+	key := hashMatchers(matchers, mint, maxt, hints)
 	if _, ok := p.selectors[key]; !ok {
-		p.selectors[key] = newSeriesSelector(p.queryable, mint, maxt, matchers)
+		p.selectors[key] = newSeriesSelector(p.queryable, mint, maxt, step, matchers, hints)
 	}
 
 	return NewFilteredSelector(p.selectors[key], NewFilter(filters))
 }
 
-func hashMatchers(matchers []*labels.Matcher, mint, maxt int64) uint64 {
+func hashMatchers(matchers []*labels.Matcher, mint, maxt int64, hints *storage.SelectHints) uint64 {
 	sb := xxhash.New()
 	for _, m := range matchers {
-		_, _ = sb.WriteString(m.Name)
-		_, _ = sb.Write(sep)
-		_, _ = sb.WriteString(strconv.Itoa(int(m.Type)))
-		_, _ = sb.Write(sep)
-		_, _ = sb.WriteString(m.Value)
-		_, _ = sb.Write(sep)
+		writeMatcher(sb, m)
 	}
-	_, _ = sb.WriteString(fmt.Sprintf("%d", mint))
-	_, _ = sb.Write(sep)
-	_, _ = sb.WriteString(fmt.Sprintf("%d", maxt))
+	writeInt64(sb, mint)
+	writeInt64(sb, maxt)
+	writeInt64(sb, hints.Step)
+	writeString(sb, hints.Func)
+	writeString(sb, strings.Join(hints.Grouping, ";"))
+	writeBool(sb, hints.By)
 
 	key := sb.Sum64()
 	return key
 }
+
+func writeMatcher(sb *xxhash.Digest, m *labels.Matcher) {
+	writeString(sb, m.Name)
```

Comment:
Do we need to write the separator here like previously? as far as I can tell the label value potentially could be the same as the label name? :thinking: